.gitmodules

```diff
@@ -0,0 +1,4 @@
+[submodule "3rdparty/optee/optee_os"]
```

Comment:
Oh I skipped right past this. We have so far avoided adding submodules to OE for several very good reasons.

1) They make using Git much more difficult for those not familiar with submodules.
2) They introduce a network based dependency for every "clean" build (like done from CI), which can cause CI failures.
3) Git uses SHA1 for its hashes, and since SHA1 is known to be quite broken, there's no actual security assurance that the submodule checkout is exactly the code the developers said it should be.
4) It adds another location to monitor for security breaches etc.

So we have opted for all other 3rdparty code to commit and redistribute (with appropriate notices of course) the code directly within our repo. Obviously #3 still applies but since it's coming from a single trusted location, it's easier to be certain about.

How opposed would your team be to following this route instead of a submodule?

/cc @CodeMonkeyLeet @mikbras

Reply:
Very opposed.  I don't consider any of those reasons to be "very good" reasons (i.e., the supposed benefit is not worth all the problems, and the benefits of submodules outweigh them).   Instead, submodules:
1) make it easier/safer to incorporate external bugfixes in submodules, since code won't diverge, which can lead to security vulnerabilities
2) encourage contributing back code in submodules rather than encouraging maintaining separate copies
3) reduce the footprint of the specific repo, making code reviews easier
4) make licensing easier since the entire OE repo can have a single license

I would strongly recommend all the other copies of 3rdparty code to be removed and replaced with submodules.

Reply:
I back Dave on this one. Additionally, our fork of OP-TEE OS picks up changes fairly often, which we would then have to replicate here.

Reply:
Do you want to teach everyone how to use submodules? 🤣 I laugh, but that is a very high cost that is dangerous to underestimate (I've been down this road several times).

Reply:
I'd be interested in hearing what problems you've had with submodules. I've found that they work just fine once you know what to look out for. We can set up a short meeting and discuss the pros and cons if you'd like.

Reply:
All of the other OSS projects (which are not Microsoft-specific by the way) I've ever worked on use submodules.  So: yes.

Reply:
I think for the basic requirements (that we can develop our project while utilizing 3rdparty code), both submodules and checking in source (either unarchived or archived) will be satisfactory. There are some advantages and disadvantages to both styles of including 3rdparty code (as detailed above), and personally I don't have a strong opinion as they both accomplish my basic requirements. I have seen both styles used in the open source world, for example mesos utilizes a similar style to what we use in OE. (There's another style too, which is where users are told that they need to download/extract/(patch)/install 3rdparty source dependences, and then they specify those extracted 3rdparty files at configure time. In my opinion that can be a nightmare for developers, unless it's a pre-built package that can be pulled through an APT/yum repo. I see this style often in open source projects, and I really find it annoying!)

Even though I find both styles to be satisfactory, I think it would be awkward to mix & match styles, where we have some 3rdparty code that comes from a submodule and some 3rdparty code that has its source checked into our repository.  If we decide on making this change, we should probably do it for all of our 3rdparty dependencies out of consistency.

Maybe we should organize a meeting to come up with a plan for the future on what style of 3rdparty references we should maintaining? What do you all think?

Reply:
I second Andy on this. It's hard to comply to SDL with an open source project, it's even harder to achieve compliance with submodules pointing to other open source projects. SDL requires every version of every 3rd party library need to be on an "approved" list. So the #1 benefit of submodules, easier/safer to incorporate external bugfixes, is not practical, considering the cost of get approval for those bug fixes.

Reply:
**What I would prefer is the option where we can keep track of the changes we make to 3rd party code, and reapply them when we upgrade the 3rd party libraries.**
 
The key point to keep in mind is that **we don't use 3rd-party code as-is**. Since the code must  run within an enclave, we have to make changes to it.
 
Some 3rd party code like mbedtls have a nice config.h which can be edited to configure it they way we want. 

Some times the changes involve directly editing code deep in the guts of the 3rd-party module (see https://github.com/microsoft/openenclave/commit/5e8ea112910f9964666a684f110c1357f47e0c91#diff-d45)

Not all changes can be upstreamed since these changes are specific to an SGX enclave.

Which of the two approaches (or some other approach) would allow us to keep track of the changes we have done to 3rd party code and **automatically** make sure the changes are reapplied when we upgrade the 3rd party modules?

The issue I see with the current approach is that when a 3rd party module is upgraded as in https://github.com/microsoft/openenclave/pull/1786/files, there are a large number of files to look at, it is not possible to figure out which of those files contain manual edits. 
I would instead prefer the PR to just contain the set of our changes plus just a reference to the libunwind version we are upgrading to.

Reply:
I don't see this as a blocking issue if checked in after we do the 0.6 release branch off master, then create an issue to discuss this in a proper forum and come to a conclusion one way or another.

Reply:
I do see adding submodules right now as a blocking issue.

Reply:
I'm in agreement that we should go with submodules as our standard way of consuming 3rdparty source code, but I think we should make a transition plan before we quickly switch to using them, or we might introduce some problems. For example, our CI process may need to do additional cleanup for its submodules between builds or you might have sequential builds interacting with each other, unpredictably influencing the outcome of a CI run. There's several other things that need consideration that Andy brought up in our discussion earlier this week, like that people will now need to do a recursive clone or a submodule update --init, or that our build system should likely warn people if the submodules haven't been initialized yet, and so on.

I imagine the change to using submodules probably needs to be done in a separate PR. What do you all think about that? Tagging @ionutbalutoiu for communication with the Cloudbase team on possibly coordinating this upcoming task

Reply:
I will drive this issue. I have created issue #1902